### ISSUE TITLE
1.2.2 raises exceptions if replica set secondary member is offline

### ISSUE THREAD

**Issue by jakolehm** [kontena/kontena#2355]
After one of the secondary replset members is stopped:

```
2017-05-23T08:14:20.000Z [1]: E, [2017-05-23T08:14:20.463041 #12] ERROR -- : Actor crashed!
2017-05-23T08:14:20.000Z [1]: Moped::Errors::PoolTimeout: ConnectionPool::PoolShuttingDownError
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:124:in `rescue in connection'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:115:in `connection'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:151:in `disconnect'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/failover/retry.rb:27:in `execute'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:200:in `rescue in ensure_connected'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:203:in `ensure_connected'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:603:in `block in flush'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:632:in `block in logging'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/activesupport-4.1.14/lib/active_support/notifications.rb:159:in `block in instrument'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/activesupport-4.1.14/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/activesupport-4.1.14/lib/active_support/notifications.rb:159:in `instrument'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/instrumentable.rb:31:in `instrument'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:631:in `logging'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:602:in `flush'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:406:in `process'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/operation/read.rb:48:in `execute'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:663:in `read'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:90:in `command'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:447:in `refresh'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/cluster.rb:188:in `block in refresh'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/cluster.rb:200:in `each'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/cluster.rb:200:in `refresh'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/cluster.rb:151:in `nodes'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/cluster.rb:246:in `with_primary'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/read_preference/primary.rb:55:in `block in with_node'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/retryable.rb:30:in `with_retry'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/read_preference/primary.rb:54:in `with_node'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/database.rb:72:in `command'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/session.rb:83:in `command'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/query.rb:309:in `block in modify'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/session.rb:374:in `with'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/query.rb:308:in `modify'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual/find_and_modify.rb:44:in `result'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual/mongo.rb:182:in `find_and_modify'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual.rb:20:in `find_and_modify'
2017-05-23T08:14:20.000Z [1]: 	/app/app/models/distributed_lock.rb:41:in `obtain_lock'
2017-05-23T08:14:20.000Z [1]: 	/app/app/models/distributed_lock.rb:20:in `with_lock'
2017-05-23T08:14:20.000Z [1]: 	/app/app/helpers/distributed_locks.rb:6:in `with_dlock'
2017-05-23T08:14:20.000Z [1]: 	/app/app/jobs/leader_elector_job.rb:31:in `elect'
2017-05-23T08:14:20.000Z [1]: 	/app/app/jobs/leader_elector_job.rb:24:in `block in perform'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
2017-05-23T08:14:20.000Z [1]: 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
2017-05-23T08:14:20.000Z [1]: I, [2017-05-23T08:14:20.464243 #12]  INFO -- LeaderElectorJob: participating leader elections
2017-05-23T08:14:20.000Z [1]: 10.81.128.78 - - [23/May/2017:08:14:20 +0000] "GET / HTTP/1.0" 200 78 0.0003
2017-05-23T08:14:22.000Z [1]: 10.81.128.36 - - [23/May/2017:08:14:22 +0000] "GET / HTTP/1.0" 200 78 0.0003
2017-05-23T08:14:22.000Z [1]: 10.81.128.53 - - [23/May/2017:08:14:22 +0000] "GET / HTTP/1.0" 200 78 0.0003
2017-05-23T08:14:22.000Z [1]: 10.81.128.78 - - [23/May/2017:08:14:22 +0000] "GET / HTTP/1.0" 200 78 0.0003
```

**Comment by SpComb** [comment]
The [`Moped::Node#connection` `raise Moped::Errors::PoolTimeout`](https://github.com/mongoid/moped/blob/v2.0.7/lib/moped/node.rb#L124) caused by a `ConnectionPool::PoolShuttingDownError` is only raised in [`ConnectionPool::TimedStack#pop`](https://github.com/mperham/connection_pool/blob/v2.2.0/lib/connection_pool/timed_stack.rb#L78), which indicates that something has called `ConnectionPool.shutdown`.

**Comment by SpComb** [comment]
The moped 2.0.x release series has a long history of issues and various fixes around these exact issues (`Moped::Node` connection pool behavior on errors). My feeling is that the moped driver (as used in Kontena 1.2) is just broken re mongo replset error handling/failover, and everyone has given up on fixing it and switched over the newer mongo driver, as used in Kontena 1.3...

**Comment by jakolehm** [comment]
Seems to be fixed in 1.3.0-dev (master branch) with Mongoid 5.2. I only got one error after I shutdown primary node (shutdown of secondaries did not produce any errors):

```
2017-05-23T11:14:27.000Z [1]: E, [2017-05-23T11:14:27.031836 #11] ERROR -- MongoPubsub: error while tailing: interrupted at shutdown (11600)
2017-05-23T11:14:27.000Z [1]: E, [2017-05-23T11:14:27.031892 #11] ERROR -- MongoPubsub: ["/usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/result.rb:256:in `validate!'", "/usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/executable.rb:37:in `block in execute'", "/usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection_pool.rb:107:in `with_connection'", "/usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/server.rb:242:in `with_connection'", "/usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/executable.rb:35:in `execute'", "/usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/cursor.rb:188:in `block in get_more'", "/usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/retryable.rb:51:in `read_with_retry'", "/usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/cursor.rb:187:in `get_more'", "/usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/cursor.rb:113:in `each'", "/usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/collection/view/iterable.rb:44:in `each'", "/usr/lib/ruby/gems/2.3.0/gems/mongoid-5.2.1/lib/mongoid/query_cache.rb:221:in `each'", "/app/app/services/mongo_pubsub.rb:155:in `block in tail!'", "/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'", "/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'", "/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:16:in `dispatch'", "/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/future.rb:18:in `block in new'", "/usr/lib/ruby/gems/2.3.0/gems/celluloid-essentials-0.20.5/lib/celluloid/internals/thread_handle.rb:14:in `block in initialize'", "/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor/system.rb:78:in `block in get_thread'", "/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:50:in `block in instantiate'"]
```
